source/hooks/hooks/d3d9hooks.cpp

```diff
@@ -959,53 +962,185 @@ namespace Hooks
 
 
 
-    struct MyDirect3D9
-    {
-        static BOOL Hook(IDirect3D9* obj)
-        {
-            BOOL rv = FALSE;
-            rv |= VTHOOKFUNC(IDirect3D9, CreateDevice);
-            rv |= HookVTable(obj, 0, (FARPROC)MyQueryInterface, (FARPROC&)QueryInterface, __FUNCTION__": QueryInterface");
-            return rv;
-        }
+    #define CALL(name) \
+        s_vtable_to_original[*reinterpret_cast<void**>(thisptr)]. ## name
 
-        static HRESULT(STDMETHODCALLTYPE *QueryInterface)(IDirect3D9* pThis, REFIID riid, void** ppvObj);
-        static HRESULT STDMETHODCALLTYPE MyQueryInterface(IDirect3D9* pThis, REFIID riid, void** ppvObj)
+    /*
+     * If this vtable was not previously hooked, this will create a new entry in the map.
+     * If this vtable was previously hooked partially, this will only hook the new entries.
+     * (for example if the Ex vtable and the non-Ex vtable start at the same spot).
+     * If this vtable was previously hooked, this will not do anything as expected.
+     * -- YaLTeR
+     */
+    #define HOOK(iface, name) \
+        HookVTable( \
+            obj, \
+            offsetof_virtual(&iface::name), \
+            reinterpret_cast<FARPROC>(My ## name), \
+            reinterpret_cast<FARPROC&>(s_vtable_to_original[*reinterpret_cast<void**>(obj)]. ## name), \
+            __FUNCTION__ ": " #name);
+
+    class MyDirect3D9
+    {
+    protected:
+        /*
+         * Map from pointers to vtables to our structs containing the original function pointers.
+         */
+        static std::map<LPVOID, MyDirect3D9> s_vtable_to_original;

```

Comment:
If you make the second type an `LPVOID`, can you move `CreateDeviceEx()` into `MyDirect3D9Ex` then? You should be able to perform `reinterpret_cast<decltype(this)>()` on it to get the correct pointer back.